src/models/BanList.ts

```diff
@@ -281,6 +286,71 @@ class BanList extends EventEmitter {
         this.emit('BanList.update', this, changes);
         return changes;
     }
+
+    /**
+     * Inform the `BanList` about a new event from the room it is modelling.
+     * @param event An event from the room the `BanList` models to inform an instance about.
+     */
+    public updateForEvent(event: { event_id: string }): void {
+        // We have to allow the batcher to emit BanList.batch because
+        // if we await in the updateForEvent method that is called by Mjolnir's sync
+        // event emitter, then by the time we start batching we will be far too late
+        // and unable to batch effectivly
+        // if you don't believe me you can test it for yourself, it is rubbish.
+        this.batcher.addToBatch(event.event_id)
+    }
 }
 
 export default BanList;
+
+/**
+ * Helper class that emits a batch event on a `BanList` when it has made a batch
+ * out of the events given to `addToBatch`.
+ */
+class UpdateBatcher {
+    private isWaiting = false;
+    private previousEventId: string|null = null;
+    private readonly waitPeriod = 200; // 200ms seems good enough.
+    private readonly maxWait = 3000; // 3s is long enough to wait while batching.
+
+    constructor(private readonly banList: BanList) {
+
+    }
+
+    /**
+     * Reset the state for the next batch.
+     */
+    private reset() {
+        this.previousEventId = null;
+        this.isWaiting = false;
+    }
+
+    /**
+     * Checks if any more events have been added to the current batch since
+     * the previous iteration, then keep waiting up to `this.maxWait`, otherwise stop
+     * and emit a batch.
+     * @param eventId The id of the first event for this batch.
+     */
+    private async checkBatch(eventId: string): Promise<void> {
+        let start = Date.now();
+        do {
+            await new Promise(resolve => setTimeout(resolve, this.waitPeriod));
+        } while ((Date.now() - start) < this.maxWait && this.previousEventId !== eventId)
+        this.reset();
+        this.banList.emit('BanList.batch', this.banList);
```

Comment:
As an academic exercise, I'm trying to understand the control flow:

This calls `Mjolnir.syncWithBanList`;

1. which calls *back* into `banList.updateList`;
   a. this reads from the room;
   b. then it updates our in-memory cache;
   c. and then it emits `BanList.update`
      - which updates the RuleServer (I think that's all?)
2. then it calls `applyServerAcls`;
3. then it calls `applyUserBans`;
4. then it calls `processRedactionQueue`.

Right?

Do I understand correctly that the main objective of this patch is to decrease how often we perform these last three calls?

Reply:
It looks to me like the PR could be a bit simpler if we were to batch `applyServerAcls` and `applyUserBans` instead of `BanList.updateForEvent` but that would break `processRedactionQueue`, right?

Reply:
>It looks to me like the PR could be a bit simpler if we were to batch applyServerAcls and applyUserBans instead of BanList.updateForEvent but that would break processRedactionQueue, right?

I don't think it would break processRedactionQueue, but it would be messier, it is what i tried originally. This can be simplified further though. 

>As an academic exercise, I'm trying to understand the control flow:

Honestly, i really don't think async/await and Promises are the right way for languages to go for this reason, but what do I know?

Reply:
> I don't think it would break processRedactionQueue, but it would be messier, it is what i tried [originally](https://github.com/matrix-org/mjolnir/blob/b7eb12997897c24d6ceeb51b303fc8595b51e67c/src/actions/policySyncBatcher.ts). What we have now can be simplified further though.

I'd prefer if it were simpler, but that can wait for a followup.

Reply:
Your initial list is right 

>async/await/Promise

This was followed up on Matrix for any on lookers